### PR TITLE
Addition of ExponentialFamily

### PR DESCRIPTION
**Pull request by vishwakftw**
This adds `ExponentialFamily`. `ExponentialFamily` uses concepts of Bregman divergence of the log normalizers of distributions to compute entropy and KL-divergences.

For distributions belonging to an exponential family, this PR adds tests to check implementations of entropy and KL-Divergence functions with both the Monte Carlo and Bregman divergences.

Joint work with @fritzo . Math checked and approved by @fritzo .

### CODE REVIEW

**Comment by fritzo** [comment]
I've checked math in: `ExponentialFamily.entropy()`, `ExponentialFamily.kl_divergence()`, and a couple of the `.natural_params` and `.log_normalizer()` implementations.

This PR is mainly for testing: it provides alternative implementations for `.entropy()` and `kl_divergence()` and tests that our two implementations agree; it thus provides higher assurance of correctness than our existing Monte Carlo tests. The implementations in this PR are not used by default in any of our existing distributions.

File: torch/distributions/dirichlet.py
Comment (apaszke) [inline torch/distributions/dirichlet.py]: Do we plan on having those methods be part of the public interface? They're doing things that aren't fully legal in autograd (rewrapping data in other Variables like this). The semantics are also unclear for me (why does the result always require grad)?

Comment (fritzo) [reply torch/distributions/dirichlet.py]: I suppose these could be private, since I don't know of an application outside of `.entropy()` and `.kl_divergence()`; however there may be future applications.

The `.natural_params` need to set `requires_grad=True` internally because `.entropy()` and `.kl_divergence` both use `grad()` under the hood, even if the original variables do not require grad. Would you recommend wrapping in `Variable(-, requires_grad=True)` only in the case that the original variables do not require grad? Or is there a better practice we can follow?

Comment (apaszke) [reply torch/distributions/dirichlet.py]: A better practice would be to return `self.concentration` and let the caller decide if they want to take grads or not

Comment (fritzo) [reply torch/distributions/dirichlet.py]: Regardless of whether the user wants to take grads, `.entropy()` and `.kl_divergence()` need to use `grad()` under the hood; that is why we are wrapping this with `requires_grad=True`. That is, even if the user does not want to compute gradients of `Dirichlet(concentration).entropy()`, the Bregman divergence implementation uses `grad()` under the hood and must use a differentiable form of `concentration`, i.e. even if the result of `.entropy()` is not differentiable wrt the user's input. See implementation of `ExponentialFamily.entropy()` for clarification.

Comment (apaszke) [reply torch/distributions/dirichlet.py]: I understand. I'm just saying that setting `requires_grad=True` should be a responsibility of the caller. If a particular implementation of `entropy()` needs it, then it needs to set the flag.

Comment (fritzo) [reply torch/distributions/dirichlet.py]: Ok, if I understand correctly then we should simply fail in the following case?
```py
dist = Dirichlet(Variable(torch.ones(2), requires_grad=False))
Dirichlet.entropy(dist)          # Succeeds.
ExponentialFamily.entropy(dist)  # Should this succeed or fail?
```
@vishwakftw I think in this case we can simply add `requires_grad=True` to all the test examples? And maybe add a check in `.entropy()` and `.kl_divergence()`:
```py
nparams = self.natural_params
if not all(p.requires_grad for p in nparams):
    raise NotImplementedError(".entropy() requires all parameters to requires_grad")
```

Comment (apaszke) [reply torch/distributions/dirichlet.py]: No, I meant this:
```python3
class ExponentialFamily(Distribution):
  def entropy(self):
    natural_params = [Variable(p, requires_grad=True) for p in self.natural_params]
    ...
```

Comment (vadimkantorov) [reply torch/distributions/dirichlet.py]: If this is frequent (setting `require_grad = True` on existing variables to enable `grad()`, how about adding a convenience method `attach()`? (that was also my suggestion in https://github.com/pytorch/pytorch/issues/1734)

**Comment by vishwakftw** [comment]
After the discussion between @apaszke and @fritzo , I have refactored the system a bit. Please have a look at the commit message for more details.

File: torch/distributions/exp_family.py
Comment (fritzo) [inline torch/distributions/exp_family.py]: I think this now needs to be a function
```py
class ExponentialFamily(Distribution):
    def log_normalizer(self, *params):
        raise NotImplementedError

class Exponential(ExponentialFamily):
    def log_normalizer(self, x):
        return -torch.log(-x)
```

File: torch/distributions/exponential.py
Comment (fritzo) [inline torch/distributions/exponential.py]: A property that returns a lambda is just a function :smile:

File: torch/distributions/exp_family.py
Comment (fritzo) [inline torch/distributions/exp_family.py]: Is it the case that `mean_carrier_measure` never depends on params? If it never depends on params, then this implementation works fine. However if it ever does depend on params, then you should change this to a function with the same signature as `log_normalizer()`.

**Review by fritzo — COMMENTED** [review]
Thanks for refactoring!

File: torch/distributions/exp_family.py
Comment (vishwakftw) [reply torch/distributions/exp_family.py]: Expected Carrier measure is independent of natural params.

File: torch/distributions/exponential.py
Comment (vishwakftw) [reply torch/distributions/exponential.py]: I realized that after I pushed, because this is redundant.

File: torch/distributions/exp_family.py
Comment (vishwakftw) [reply torch/distributions/exp_family.py]: Does the function signature need to be the same? I changed it without changing the signature in `ExponentialFamily` and it still seems to be fine.

File: torch/distributions/exp_family.py
Comment (vishwakftw) [reply torch/distributions/exp_family.py]: At least for the distributions we have now, `mean_carrier_measure` is independent of `natural_params`.

File: torch/distributions/exp_family.py
Comment (fritzo) [reply torch/distributions/exp_family.py]: I think `log_normalizer(self, *natural_params)` is not needed but would clarify how it should be implemented and would serve as documentation for developers.

**Comment by fritzo** [comment]
@vishwakftw I kind of like Adam's suggestion of making `._natural_params`, `._log_normalizer()` and `.mean_carrier_measure` private, and keeping only `.entropy()` and `.kl_divergence()` public. What do you think? For the same reason, should we move `.kl_divergence()` from a method to a registered KL implementation?
```py
@register_kl(ExponentialFamily, ExponentialFamily)
def _kl_expfamily_expfamily(p, q):
    if type(p) is not type(q):
        raise NotImplementedError
    ...your implementation here...
```
This would prevent users from accidentally using `ExponentialDistribution.kl_divergence()` outside of testing code; in tests we could directly reference `_kl_expfamily_expfamily()`. WDYT?

**Comment by vishwakftw** [comment]
I think this is fine w.r.t. KL-Divergences, but what can we do about `.entropy`?

File: torch/distributions/exp_family.py
Comment (apaszke) [reply torch/distributions/exp_family.py]: Is it expected that this function has a different signature across distributions? Does the number of arguments match the number of natural params?

**Review by apaszke — COMMENTED** [review]
This looks good. I have three main questions:
1. What's the signature of `log_normalizer` (see comment inline)
2. Are you sure that this API will be stable, or is it an experimental thing mainly for test purposes? If e.g. there exist distributions where the way `mean_carrier_measure` is not implemented this way, we might want to change the signature.
3. Why does `natural_params` return tensors and not variables?

**Comment by vishwakftw** [comment]
@apaszke Thanks for the questions:
1. The function signature for `log_normalizer` is `(self, *natural_params)`. Yes, the number of arguments is exactly the number of natural parameters characterizing the distribution. I will modify default implementation in `ExponentialFamily` to match this, as suggested by @fritzo 
2. Theoretically, the mean carrier measure could be a function of natural parameters. But so far, in all the distributions we have seen, this is not the case, and has either been 0 or doesn't have a closed form (which is when we leave it as it is).
3. `natural_params` could return the same type as the actual parameters for the distribution, but this will cause some case-checking and rewrapping in the entropy and KL-divergences, when we have to use the AD system for compute Bregman Divergences.

Let me know your views about this, and clarifications if any. If this is clear, then I will proceed to modify the default function signature. If you have anything you would like me to add in the code, I will make that addition as well. I hope I have addressed your questions well.

File: torch/distributions/kl.py
Comment (fritzo) [inline torch/distributions/kl.py]: I believe this should be a `NotImplementedError`: it is a shortcoming of our implementation rather than a user error due to poorly constructed data.

**Comment by apaszke** [comment]
Thanks @vishwakftw. 
2. But even if that will be the case, it should be still possible to compute the mean carrier measure, because it could just be a property, right?
3. I still think that returning Variables is safer, and we will be able to remove most checking code once Variables and Tensors are merged. You could even just have an assert that checks that args aren't tensors.

**Comment by vishwakftw** [comment]
@apaszke Following your suggestions, I have generalized the code to take only `Variable` as input. I have not added any asserts for that however. 

Regarding the mean carrier measure, existing literature so far suggests that it is either 0 or intractable to compute. Maybe @fritzo and @alicanb can weigh in on this.

File: torch/distributions/exponential.py
Comment (apaszke) [inline torch/distributions/exponential.py]: Can you simplify those methods to one liners like `(-self.rate,)`? Also, do we really want them to be lazy properties? They will hold on to their results, consuming more memory than necessary.

**Review by apaszke — COMMENTED** [review]
I have some general suggestions for how `natural_params` are implemented (only applies in those extremely simple cases, e.g. `Normal` is fine). I'd still make them regular `@property` in all classes to save memory.

Once this is addressed and conflicts are fixed, it should be good to merge!

**Comment by fritzo** [comment]
@vishwakftw I still like Adam's suggestion of making `._natural_params`, `._log_normalizer()` and `._mean_carrier_measure` private, since they are only used to implement generic methods. Mostly I don't want to confuse users of `Normal` with internal-only methods, since that detracts attention from important methods like `.rsample()` and `.entropy()`. Does that sound reasonable?

**Comment by vishwakftw** [comment]
@fritzo I think that is a good idea. Especially, because of the fact that natural parameters will be occasionally used, I think it would be better to abstract it away.

We can come with a similar logic as done for `kl_divergence` for these methods as well. Please let me know what you think.

**Comment by fritzo** [comment]
@vishwakftw can we just add underscores to the three internal `ExponentialFamily` methods?

**Review by fritzo — APPROVED** [review]
Looks great!

File: test/test_distributions.py
Comment (apaszke) [inline test/test_distributions.py]: any reason why these two are `variable` instead of `Variable`?

File: test/test_distributions.py
Comment (apaszke) [inline test/test_distributions.py]: Why `ValueError`?

File: torch/distributions/kl.py
Comment (apaszke) [inline torch/distributions/kl.py]: @fritzo added a helper to compute this sum more efficiently in another PR. I'll merge that one first, and then please rebase and update this code

**Review by apaszke — COMMENTED** [review]
Blocked on #4937

File: test/test_distributions.py
Comment (vishwakftw) [reply test/test_distributions.py]: It is a constructor difference, right? I thought it would be unnecessary to wrap `[1.0, 1.0]` with a `Tensor` and then with `Variable`.

However, to maintain consistency, I am changing this. Thanks for pointing this out.

File: test/test_distributions.py
Comment (vishwakftw) [reply test/test_distributions.py]: Sorry about that. That was there when the `.entropy` in `ExponentialFamily` returned a `ValueError`. I have removed it now. Thank you.

File: torch/distributions/kl.py
Comment (vishwakftw) [reply torch/distributions/kl.py]: Sure. Thank you.

**Comment by vishwakftw** [comment]
@apaszke I have modified the kl computation using @fritzo's helper function.

File: torch/distributions/kl.py
Comment (fritzo) [inline torch/distributions/kl.py]: Nice, thanks for correctly handling the multivariate case.

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Thanks @vishwakftw!

**Pull request merged**
